File: c/chapitre_7/exo72Affiche.c
/*
crire un programme saisit.c qui saisit un int au clavier, et l’enregistre
dans un ﬁchier /tmp/entier.txt. Écrire un programme affiche.c qui attend (avec sleep) un
signal utilisateur du programme saisit.c. Lorsque l’entier a été saisi, le programme affiche.c
aﬃche la valeur de l’entier.
*/

#define _POSIX_C_SOURCE 200112L

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

/* Global pour accès dans le handler */
int recu = 0;

void handler(int sig)
{
    recu = 1;
}

int main()
{
    FILE *f;
    int valeur;
    struct sigaction sa ;

    // Confid signal
    sa.sa_handler = handler;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction(SIGUSR1, &sa, NULL);

    printf("Mon PID : %d\n", getpid());
    printf("En attente du signal...\n");

    /* Attente active */
    while (!recu)
    {
        sleep(1);
    }

    f = fopen("/tmp/entier.txt", "r");
    if (f)
    {
        fscanf(f, "%d", &valeur);
        printf("Valeur lue : %d\n", valeur);
        fclose(f);
    }

    return 0;
}

File: python/chapitre_7/exo72Affiche.py
"""
crire un programme saisit.c qui saisit un int au clavier, et l'enregistre
dans un ﬁchier /tmp/entier.txt. Écrire un programme affiche.c qui attend (avec sleep) un
signal utilisateur du programme saisit.c. Lorsque l'entier a été saisi, le programme affiche.c
aﬃche la valeur de l'entier.
"""

import os
import sys
import signal
import time

# Global pour accès dans le handler
recu = 0

def handler(sig, frame):
    global recu
    recu = 1

def main():
    global recu

    # Config signal
    signal.signal(signal.SIGUSR1, handler)

    print(f"Mon PID : {os.getpid()}")
    print("En attente du signal...")

    # Attente active
    while not recu:
        time.sleep(1)

    try:
        with open("/tmp/entier.txt", "r") as f:
            valeur = int(f.read())
            print(f"Valeur lue : {valeur}")
    except Exception:
        pass

    return 0

main()